Patch/pinned init v3

BennoLossin opened the pull request.

metaspace commented on rust/macros/pin_data.rs:
```diff
@@ -0,0 +1,79 @@
+
+pub(crate) fn pin_data(args: TokenStream, input: TokenStream) -> TokenStream {
+    // This proc-macro only does some pre-parsing and then delegates the actual parsing to
+    // `kernel::_pin_data!`.
```
There is a typo here, should be `__pin_data`, right?

BennoLossin replied:
That is correct, good catch!

ojeda commented on rust/macros/pin_data.rs:
```diff
@@ -0,0 +1,79 @@
+    // The current level of `<`.
+    let mut nesting = 0;
+    let mut toks = input.into_iter();
+    // if we are at the beginning of a generic parameter
```
```suggestion
    // If we are at the beginning of a generic parameter.
```

ojeda commented on rust/macros/lib.rs:
````diff
@@ -166,3 +170,79 @@ pub fn vtable(attr: TokenStream, ts: TokenStream) -> TokenStream {
+/// ```
+///
+/// [`pin_init!`]: ../kernel/macro.pin_init.html
+//  ^ cannot use direct link, since `kernel` is not a dependency of `macros`
````
```suggestion
//  ^ cannot use direct link, since `kernel` is not a dependency of `macros`.
```

ojeda commented on rust/macros/lib.rs:
```diff
@@ -166,3 +170,79 @@ pub fn vtable(attr: TokenStream, ts: TokenStream) -> TokenStream {
+/// Place this macro on a struct definition and then `#[pin]` in front of the attributes of each
+/// field you want to structurally pin.
+///
+/// This macro enables the use of the [`pin_init!`] macro. When pinned-initializing a `struct`,
```
Should it be "pin-initializing"?

ojeda commented on rust/kernel/sync/arc.rs:
```diff
@@ -163,6 +167,28 @@ impl<T> Arc<T> {
+
+    /// Use the given initializer to in-place initialize a `T`.
+    ///
+    /// This is equivalent to [`pin_init`], since an `Arc` is always pinned.
```
Can the `Arc` be an intra-doc link?

ojeda commented on rust/kernel/init/macros.rs:
```diff
@@ -0,0 +1,481 @@
+//! This architecture has been chosen because the kernel does not yet have access to `syn` which
+//! would make matters a lot easier for implementing these as proc-macros.
+
+/// This macro creates a `unsafe impl<...> PinnedDrop for $type` block.
```
```suggestion
/// Creates a `unsafe impl<...> PinnedDrop for $type` block.
```

ojeda commented on rust/kernel/init/macros.rs:
```diff
@@ -0,0 +1,481 @@
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This module provides the macros that actually implement the proc-macros `pin_data` and
```
Typically we have a single line with the title ("short description"). It does not matter much since this is `doc(hidden)`, but still, it would be nice to have it.

BennoLossin commented:
for anyone wanting to see the individual commits, see https://github.com/y86-dev/linux/tree/patch/pinned-init-v3-not-squashed

BennoLossin commented on rust/kernel/lib.rs:
```diff
@@ -16,7 +16,9 @@
 #![feature(dispatch_from_dyn)]
+#![feature(explicit_generic_args_with_impl_trait)]
 #![feature(generic_associated_types)]
+#![feature(new_uninit)]
```
should these feature enables be in their own commits?

ojeda replied:
Don't worry about it -- Lina has a patch enabling it too, so I can pick that one first: https://lore.kernel.org/rust-for-linux/20230224-rust-new_uninit-v1-1-c951443d9e26@asahilina.net/

BennoLossin commented on scripts/Makefile.build:
```diff
@@ -277,7 +277,7 @@ $(obj)/%.lst: $(src)/%.c FORCE
 # ---------------------------------------------------------------------------
 
-rust_allowed_features := core_ffi_c
+rust_allowed_features := core_ffi_c,explicit_generic_args_with_impl_trait
```
Similarly this?

This feature was stabilized in 1.63, so this might not be here for long.

ojeda replied:
Don't worry about it -- Lina has a patch enabling it too, so I can pick that one first: https://lore.kernel.org/rust-for-linux/20230224-rust-new_uninit-v1-1-c951443d9e26@asahilina.net/

ojeda replied:
No worries either, it will go away in the merge cycle.

metaspace commented on rust/kernel/init.rs:
```diff
@@ -0,0 +1,1337 @@
+        @munch_fields($field:ident $(: $val:expr)?, $($rest:tt)*),
+    ) => {
+        $(let $field = $val;)?
+        // Call the initializer.
```
Should this be "Initialize the field" instead maybe? Since no initializer is actually called in this case?

metaspace commented:
I think we should split up `rust: add pin-init API` in multiple patches. I realize that many of the types used by the macros are not meant to be used by consumers of this API, but I think this would be easier to review if the types used by the macros were added on their own, along with examples of intended use. The macros could then be added in a separate patch as a means to automate the examples.

BennoLossin commented:
> I think we should split up `rust: add pin-init API` in multiple patches. I realize that many of the types used by the macros are not meant to be used by consumers of this API, but I think this would be easier to review if the types used by the macros were added on their own, along with examples of intended use. The macros could then be added in a separate patch as a means to automate the examples.

Not sure if I understand correctly, are you talking about the types inside of `__internal.rs`, or the other types as well? What do you mean by examples of intended use? The `__internal.rs` types mostly serve a single purpose and are only intended to be used inside of the macros. I could give some macro expanded code as examples of how the macro uses these internal types if that is what you mean, but I dont know where to put them for the patches.

metaspace commented:
> Not sure if I understand correctly, are you talking about the types inside of `__internal.rs`, or the other types as well? What do you mean by examples of intended use? The `__internal.rs` types mostly serve a single purpose and are only intended to be used inside of the macros. I could give some macro expanded code as examples of how the macro uses these internal types if that is what you mean, but I dont know where to put them for the patches.

I think the `rust: add pin-init API` commit is too big. For me it was difficult to digest, so I think a bit more help to the reader would be nice. I would suggest the following:

 * Splitting the commit in at least two: One part that supports `pin_init_from_closure` and one part that adds macros that create the initializer
 * Even though `__internal` is not intended to be used by consumers of the `init` API, it is a public module that is going to be accessible to the user, right? So the user could write something like the expanded macro text, calling into `init::__internal`. Adding `__internal` in one commit along with examples of how the expanded macros will use it would be a major help in understanding the macros. The examples could be delivered in the module documentation.
 * I would suggest splitting in even more pieces, like one adding `init!` and one adding `pin_init!`
 * The commit message is nice. I would love to have that as module documentation. I think the commit message is better than the module level intro.
 * It was difficult for me to parse the macros. I am not a macro expert, I had to look up quite a bit of mechanics to follow along. One thing that would have been a great help is having expanded examples in the documentation. It was difficult for me to read the macros without knowing where it was trying to end up. I know there are tools for looking at expanded macros, but I still think some small examples in the documentation would be nice.
 * I did not immediately understand some of the traits that are used to help type inference. It would be great to document why and how this works in the expanded macro examples.

I hope this makes more sense :) I am not sure who is going to review this in the end, but if it is people that are not long time Rust experts, we need to give them all the help we can think of.

nbdd0121 commented:
Types used only internally by the macro should *not* be a separate commit. Each logical change should be a commit, no matter how big or small it is.

Splitting the commit into a core `PinInit/Init/StackInit/pin_init_from_closure` commit and a macro commit, though, does make sense.

ojeda commented on rust/kernel/init/macros.rs:
```diff
@@ -0,0 +1,503 @@
+        impl<$($impl_generics)*> MustNotImplDrop for $name<$($ty_generics)*>
+        where $($whr)* {}
+        // We also take care to prevent users from writing a useless PinnedDrop implementation.
+        // They might implement PinnedDrop correctly for the struct, but forget to give
```
```suggestion
        // We also take care to prevent users from writing a useless `PinnedDrop` implementation.
        // They might implement `PinnedDrop` correctly for the struct, but forget to give
```

ojeda commented on rust/kernel/init/macros.rs:
```diff
@@ -0,0 +1,503 @@
+            UselessPinnedDropImpl_you_need_to_specify_PinnedDrop for $name<$($ty_generics)*>
+        where $($whr)* {}
+    };
+    // When `PinnedDrop` was specified we just implement drop and delegate.
```
```suggestion
    // When `PinnedDrop` was specified we just implement `Drop` and delegate.
```

ojeda commented on rust/kernel/init/macros.rs:
```diff
@@ -0,0 +1,503 @@
+        where $($whr)*
+        {
+            fn drop(&mut self) {
+                // SAFETY: since this is a destructor, `self` will not move after this function
```
```suggestion
                // SAFETY: Since this is a destructor, `self` will not move after this function
```

ojeda commented on rust/kernel/init/macros.rs:
```diff
@@ -0,0 +1,503 @@
+                // SAFETY: since this is a destructor, `self` will not move after this function
+                // terminates, since it is inaccessible.
+                let pinned = unsafe { ::core::pin::Pin::new_unchecked(self) };
+                // SAFETY: since this is a drop function, we can create this token to call the
```
```suggestion
                // SAFETY: Since this is a drop function, we can create this token to call the
```

ojeda commented:
> Each logical change should be a commit, no matter how big or small it is.

This is true, and it simplifies reverting changes later and so on. However, big patches may be split into smaller ones to make things more understandable for reviewers, even if they could also be understood to go together as a unit too, as long as the series does not break anything after each step (so fairly easy when introducing something new).

Having said that, 2k lines is still within reason (even if already within the top percentile) and several versions have been in the list for a while, but if you (@y86-dev) can do a logical split somewhere that makes it easier to understand (taking into account the commit messages too, especially given the one you have there), please go for it for v4 since some people will appreciate it (not just now, but in the future too).

ojeda commented:
GitHub's "Compare" button says `from_value` is gone.

Also "rust: add pin-init API items to prelude" -> "rust: prelude: add pin-init API items".

BennoLossin commented:
> GitHub's "Compare" button says `from_value` is gone.
> 
> Also "rust: add pin-init API items to prelude" -> "rust: prelude: add pin-init API items".

Yeah that was replaced by `impl Init<T> for T`, had forgotten to remove it when I added that. Also some other things are not completely done, still rebasing stuff.

ojeda commented:
I saw it (`from_value`) referenced from intra-doc links too.

BennoLossin commented:
@metaspace I added a macro expansion example, let me know what you think of it.

Not sure if the `__internal.rs` docs is the best place for it, since that is `#[doc(hidden)]`, maybe I could create a `doc` module just for that purpose? Also not sure into which commit this should go, since `PinnedDrop` is now split from `pin_init!`.

metaspace commented:
> @metaspace I added a macro expansion example, let me know what you think of it.
> 
> Not sure if the `__internal.rs` docs is the best place for it, since that is `#[doc(hidden)]`, maybe I could create a `doc` module just for that purpose? Also not sure into which commit this should go, since `PinnedDrop` is now split from `pin_init!`.

I think this is really helpful. I think having it where you put it is fine. It's not something a user would read anyway.

ojeda commented on rust/kernel/init/macros.rs:
```diff
@@ -0,0 +1,964 @@
+//!
+//! This example includes the most common and important features of the pin-init API.
+//!
+//! Below you can find induvidual section about the different macro invocations. Here are some
```
```suggestion
//! Below you can find individual section about the different macro invocations. Here are some
```

fbq commented on rust/kernel/sync/arc.rs:
```diff
@@ -540,6 +540,31 @@ impl<T> UniqueArc<MaybeUninit<T>> {
+        match unsafe { init.__pinned_init(self.as_mut_ptr()) } {
+            // SAFETY: Initialization completed successfully and we can pin the value, since this
+            // is an owning pointer.
+            Ok(()) => Ok(unsafe { Pin::new_unchecked(self.assume_init()) }),
```
Maybe add a `UniqueArc::into_pin`, similar to `Box::into_pin`?

BennoLossin replied:
Oh I did not see that we already have `From<UniqueArc<T>> for Pin<UniqueArc<T>>`, I can just use that.

BennoLossin commented:
continued at #991
